### PR TITLE
Fixes #9187 - Overrideable empty puppet class parameters do not validate their defult value

### PR DESCRIPTION
**Pull request by orrabin**

### CODE REVIEW

**Comment by domcleal** [comment]
Just to check - and I think the answer's, yes - while use_puppet_default is true, I expect to be able to update the value (e.g. over API) without validation, and then to set it to false again, I need to fix the value at the same time?

**Comment by orrabin** [comment]
@domcleal yes
it should also work when you change a parameter to boolean - if you keep the default value as whatever it was before but `use_puppet_default` is true the puppet class will be saved correctly and if `use_puppet_default` is turned to false you will get a validation message.

File: app/models/lookup_key.rb [hunk 1]
```diff
@@ -30,7 +30,7 @@ def audit_class
                                 :reject_if => lambda { |a| a[:value].blank? && (a[:use_puppet_default].nil? || a[:use_puppet_default] == "0")},
                                 :allow_destroy => true
 
-  before_validation :validate_and_cast_default_value
+  before_validation :validate_and_cast_default_value, :unless => Proc.new{|p| p[:use_puppet_default]}
```
Comment (domcleal) [inline app/models/lookup_key.rb]: Nitpick, we don't usually use `[]` to access attributes, so `p.use_puppet_default?`

**Comment by domcleal** [comment]
Ok, that's fine.  Should the same check be in lookup_value?

**Comment by orrabin** [comment]
@domcleal thanks, I missed that. It's covered in the UI but you can still get an error from the API when updating a lookup_value, fixing now.

File: test/unit/lookup_key_test.rb [hunk 2]
```diff
@@ -228,4 +228,18 @@ def test_value_should_not_be_changed
     refute key.valid?
     assert_include key.errors.keys, :default_value
   end
+
+  test "default_value is not validated if use_puppet_default is true" do
```
Comment (dLobatog) [inline test/unit/lookup_key_test.rb]: Nitpick, code across all 4 tests is nearly identical. Extract to a helper?

**Comment by dLobatog** [comment]
:+1: for me, seems to work well.

**Comment by orrabin** [comment]
@eLobato I added a helper
suggestions for a better name?

**Comment by domcleal** [comment]
I think it can simply be a factory rather than a special method in test_helper.  (Got to admit, I didn't really think it deserved factoring out..)

**Comment by orrabin** [comment]
I could just create two setups, one for lookup_key and one for lookup_value just for these tests, would that be a better solution?

**Comment by domcleal** [comment]
@orrabin yeah, fine by me

**Comment by domcleal** [comment]
Merged as 92eb3d7, thanks @orrabin.

**Pull request closed without merging**